Encapsulate other scripts

dgrammatiko opened the pull request:
- The script needs to encapsulate `tippy.js`
- Tippy has a dependency on popper.js. This either will be bundled here as in this PR (advantage: no conflicts ever) or will be declared in the rollup config as external dependency and will load the script from the bootstrap folder in the CMS (disadvantages: the script is only usable in Joomla, possible conflicts)

Also the dist needs to serve ESM files. UMD was a thing 10 years ago...


@Fedik @bembelimen

Fedik commented:
The idea is to simplify development, you include ../src/..js and all works.
The tippy included separately because it is a module only for Node Environment, and not work in browser Environment.
I pack it in UMD in another PR.

> Also the dist needs to serve ESM files. UMD was a thing 10 years ago...

Yes and no. How do you suggest to use it in Joomla?
Serve ESM files not priority for now (but can do that of course), as I said it already works just including ../src/..js, that how all examples works.

dgrammatiko commented:
> The tippy included separately because it is a module only for Node Environment, and not work in browser Environment.

No, it works fine with this PR

> The idea is to simplify development, you include ../src/..js and all works.

The DX is totally broken here. The docs have everywhere hardcoded the released version of the script that is in development? Apart from the fact that someone has to update all the HTML files (no abstraction) the usage of CDN for the script in development is a really bad idea. The correct thing was to add a simple script that copies the dist files to doc and change the path in the script import. Then run it with a github action on every merge. You get both a working dev env and really maintainable docs.

> Yes and no. How do you suggest to use it in Joomla?

The problem is that Joomla already has a copy of Popper.js (an ESM for Bootstrap). If you keep the output here as a UMD then you need to have a UMD copy of Popper.js in Joomla. So basically I see 2 options:

- Encapsulate everything in one bundle (the current state of this PR)
- Abstract popper.js as an external script in the rollup config (eg like we do it for bootstrap in Joomla) and adjust also the bs5 build step in Joomla, so there is one copy shared between them.

Shipping UMD is wrong for anything new in 2022 (in a few days) but if there's no interest then you can go ahead and close this PR and do it like it's 2010.

Fedik commented:
> No, it works fine with this PR

You did not get what I wrote ;)
I meant that the tippy is broken on its own, sure it will work if we bundle it.

> Encapsulate everything in one bundle (the current state of this PR)

Yes, and then? you cannot use module in Joomla. I mean **Real Module**.
I have bundled it in dist #49 but in UMD.

> Shipping UMD is wrong for anything new

Nothing wrong with that, on some CMS you may not be able to use `import` in the same reason as we cannot do it in Joomla.
Only way is to include the script manually.

> No interest then you can go ahead and close this PR and do it like it's 2010.

As I said we can ship ESM (here your PR is okay), but we should not remove UMD (here your PR is bad).
The JavaScript from 1995 and we still use it, I see nothing wrong with that :wink:

dgrammatiko commented:
> Yes, and then? you cannot use module in Joomla. I mean Real Module.

We are using real modules for many things, Bootstrap, Custom Elements, etc

> As I said we can ship ESM (here your PR is okay), but we should not remove UMD (here your PR is bad).

Added the umd files back. Also added a simple script for the docs deployment (it needs to be called on a github action that will do the deploy [basically just run the script])
Also added a version of the files that have the popper as an external import. This needs to be replaced in the Joomla build tools (a simple replacement)

brianteeman commented:
>  the usage of CDN for the script in development is a really bad idea.

I did that for the docs as they were also used for the github.io site and you cant load the scripts from outside of the root of that webspace. I see that your docs.js script resolves that

Fedik commented:
> We are using real modules for many things, Bootstrap, Custom Elements, etc

Yes and no. Loading script with type=module does not make it a Real module, it just a script that loaded as module.
(well, technically it a Module but in my understanding it not a Real module.)
With Real Module I mean the module that `export` some own features, and it can be imported.
Only some bootstrap components are used as module.

If you need to use this instead of export:
```javascript
window.bootstrap = window.bootstrap || {};
```
Then for sure it is not a Real Module :wink: 

hmhm, in Joomla we will have setup script, probably we can do import there.
Kind of `import Jooa11y from '../../../vendor/jooa11y/jooa11y.js'`, But I not very like such hardcoding.

Fedik commented on docs/pages/script.js:
```diff
@@ -0,0 +1,10 @@
+Lang.addI18n(EngSetting.strings);
+
+// Instantiate
+const checker = new Jooa11y(EngSetting.options);
```
Please use window load event, as it was in original code.
The scan should be run only when everything is loaded.

dgrammatiko replied:
Ok, although type=module is like defer, eg document is already loaded, no need for an event

Fedik replied:
document is loaded but images are not, also the document not fully rendered

dgrammatiko replied:
Done

Fedik replied:
thanks!

Fedik commented on src/js/jooa11y.js:
```diff
@@ -1,3 +1,6 @@
+import tippy from 'tippy.js';
```
Please remove this from here, it is wrong import and prevent use `jooa11y.js` as module in browser.
You can prepend it in the build script, as I did in my PR.

dgrammatiko replied:
> it is wrong import and prevent use jooa11y.js as module in browser.

I'm not sure why you think this wrong tippy has clearly an ESM export: https://github.com/atomiks/tippyjs/blob/60aee16c934fc861569f84a37c6a1a7f42a33fa0/package.json#L6

Javascript doesn't care about node/browser as long as the dependencies exist in the context of the environment in the script is used. Also, this doesn't produce some errors so I'll take it as "it works as it was intended"

Fedik replied:
Or try change it to `import tippy from './tippy.js';`
But need to test

Fedik replied:
> Javascript doesn't care about node/browser

It care much, for Browser it is not supported import, it works only for Node.

Check https://stackoverflow.com/a/66920868/1000711
And linked issue there:

> It wouldn't work in pure Javascript because so-called “bare” module specifiers are currently not supported in ECMAScript modules specification. Module specifiers must be full URLs, or relative URLs starting with /, ./, or ../.

dgrammatiko replied:
> It care much, for Browser it is not supported import, it works only for Node.

You're confusing direct import from the source with bundling using a bundler. We are not importing tippy in the bowser from the source. We import it in the scr of the script which will then be bundled by rollup (fixing the `Uncaught TypeError: Failed to resolve module specifier "@popperjs/core"`). Check the final script (umd or esm, there are no imports)

Fedik replied:
Yes I know, but I want `jooa11y.js` to be a perfect module :wink: 
So someone can import it and use in Browser. And this import breaks whole thing.
It like 'tippy.esm.js' that even not possible to use without another compilation, because `'@popperjs/core'` :hankey:

And yes, I know we will have a bundle in dist, but I still want it to be workable even in source :wink:

dgrammatiko replied:
> So someone can import it and use in Browser. And this import breaks whole thing.

It is. Check it in your browser, also the script.js imports the `jooa11y.js` and works as expected. Everything that is needed is bundled, similar to the choices.js that bundles preact (there's no external dependency for preact in choices and similarly there's no external dependency for tippy or popper, everything that is needed is already encapsulated here)

dgrammatiko replied:
> And yes, I know we will have a bundle in dist, but I still want it to be workable even in source

Most of the ESM modules that have dependencies don't work out of the box in the browser unless the dependency is pointing to a URL. This is the reason that most scripts need a bundler with a node resolve plugin to fix those paths for the browser.

Fedik replied:
well, okay, leave it as it is, 
yea, this tippy still useless as ESM module in browser, so it not really possible to do what I want :smile:

Fedik commented:
I am a bit confused now with bunch of esm in dist :smile: 

I think we just need:
`esm` bundle `joomla-a11y-checker.esm.js`
`esm` lang `lang/en.js`
`umd` bundle `joomla-a11y-checker.umd.js`
`umd` lang `lang/en.umd.js`

dgrammatiko commented:
> hmhm, in Joomla we will have setup script, probably we can do import there.
> Kind of import Jooa11y from '../../../vendor/jooa11y/jooa11y.js', But I not very like such hardcoding.

That's what I meant, we will deal with it properly in the build tools...
ESM modules will only be possible in their full extend once Import maps are supported everywhere and you manage to add a factory in the Web Assets 😉
 
> If you need to use this instead of export:
> Then for sure it is not a Real Module 

Yes but Bootstrap is not our code, so that's other people's problem 😀

> I am a bit confused now with bunch of esm in dist

So 
- `joomla-a11y-checker.esm.min.js` is for usage in Joomla (has an external dependency for popper that needs to be fixed in the build tools)
- `joomla-a11y-checker.standalone.esm.min.js` probably this doesn't make any sense
- `joomla-a11y-checker.standalone.umd.min.js` A script that can be used in any context just by using a simple script tag (no type=module)

I will adjust the rollup config to skip the extra files

Fedik commented:
> is for usage in Joomla (has an external dependency for popper that needs to be fixed in the build tools)

no no, it should not depend from Joomla popper as module,
or we will end up in dependency hell,
just include all in bundle

dgrammatiko commented:
> just include all in bundle

Yeah that was my first thought as well

brianteeman commented:
I have tested this PR and it seems ok to the limit of my understanding of js

Fedik commented:
@brianteeman I think you can merge it if you like :)
